kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt

```diff
@@ -0,0 +1,205 @@
+                    )
+
+                else ->
+                    throw org.axonframework.serialization.SerializationException("Cannot serialize type $type to representation $expectedRepresentation. String and JsonElement are supported.")
+            }
+        } catch (ex: SerializationException) {
+            throw org.axonframework.serialization.SerializationException("Cannot serialize type ${value?.javaClass?.name} to representation $expectedRepresentation.", ex)
+        }
+    }
+
+    override fun <T> canSerializeTo(expectedRepresentation: Class<T>): Boolean =
+        expectedRepresentation == String::class.java ||
+                expectedRepresentation == JsonElement::class.java
+
+    override fun <S, T> deserialize(serializedObject: SerializedObject<S>?): T? {
+        try {
+            if (serializedObject == null) {
+                return null
+            }
+
+            if (serializedObject.type == SerializedType.emptyType()) {
+                return null
+            }
+
+            val serializer: KSerializer<T> = serializedObject.serializer()
+            return when {
+                serializedObject.contentType.isAssignableFrom(String::class.java) ->
+                    json.decodeFromString(serializer, serializedObject.data as String)
+
+                serializedObject.contentType.isAssignableFrom(JsonElement::class.java) ->
+                    json.decodeFromJsonElement(serializer, serializedObject.data as JsonElement)
+
+                else ->
+                    throw org.axonframework.serialization.SerializationException("Cannot deserialize from content type ${serializedObject.contentType} to type ${serializedObject.type}. String and JsonElement are supported.")
+            }
+        } catch (ex: SerializationException) {
+            throw org.axonframework.serialization.SerializationException(
+                "Could not deserialize from content type ${serializedObject?.contentType} to type ${serializedObject?.type}",
+                ex
+            )
+        }
+    }
+
+    private fun <S, T> SerializedObject<S>.serializer(): KSerializer<T> =
+        classForType(type).serializer() as KSerializer<T>
+
+    /**
+     * When a type is compiled by the Kotlin compiler extension, a companion object
+     * is created which contains a method `serializer()`. This method should be called
+     * to get the serializer of the class.
+     *
+     * In a 'normal' serialization environment, you would call the MyClass.serializer()
+     * method directly. Here we are in a generic setting, and need reflection to call
+     * the method.
+     *
+     * This method caches the reflection mapping from class to serializer for efficiency.
+     */
+    private fun <T> Class<T>.serializer(): KSerializer<T> =
+        serializerCache.computeIfAbsent(this) {
+            // Class<T>: T must be non-null
+            val kClass = (this as Class<Any>).kotlin
+
+            val companion = kClass.companionObject
```

Comment:
I assume that the hard requirement on the `@Serializable` annotation makes it pretty tough to use this serializer for Axon objects, like the `TrackingToken`.
I've skimmed the Kotlin Serializer documentation somewhat but didn't spot a solution for this just yet.
Do you, perchance, have a solution in mind for this, @hiddewie?

Reply:
This is the toughest question :)

As far as I can see, there are two solutions:
1. Annotate the serializable types in Axon core with `@Serializable`. This adds the compile time requirement of having the Kotlin compiler plugin dependency in the build process. That may or may not be OK. Although, there are also Jackson annotations in the source code already. In runtime there is no Kotlin requirement.
2. Create handwritten serializers for every class that needs to be serializable. This is quite some (cumbersome, error prone) work, depending on the number of serializable classes in the Axon core. The serializer code can be maintained in this repository.

To illustrate option 2, I added a small example and custom serializer with test.

The third option is abandoning this approach because the Kotlin serialization is not wanted in the core framework, and it is not worth the effort to maintain the custom serializers.

Reply:
(After thinking about option 1 a bit more: I have not tested adding `@Serializable` on a Java class and then running the Kotlin compiler plugin on it. Theoretically it should work but in practice I don't know)

Reply:
I tried a few things to test how the interoperability of Java classes with Kotlin annotations, and the Kotlin annotation processors work together.

What I did:
- Complie Axon framewok locally (4.6.0-SNAPHSOT)
- Add the Maven Kotlin compiler plugin which would compile the Kotlin annotations into `serializer` functions, see instructions here https://github.com/Kotlin/kotlinx.serialization#maven
- Add `@kotlinx.serialization.Serializable` to ConfigToken

What I saw:
- The annotation gets added (just like any other annotation) in the classfile
- No serializer static method is generated.

What I did:
- Copy the Java class and copy it to this repository, with the .java extension. 
- Let the Kotlin compiler, with the Kotlin serialization compiler plugin comple the .java class as well

What I saw:
- The annotation gets added in the classfile
- Still no serializer static method  is generated.

What I did:
- Let IntelliJ convert the .java file to a .kt Kotlin file (very ugly)

What I saw:
- The annotation gets added in the classfile
- A serializer method is generated.

So my conclusion is that the Kotlin serialization compiler plugin only works for Kotlin source files, and that the `.serializer()` static method cannot be generated for Java source files. That is very unfortunate, I might open an issue with the Kotlin Serialization team to ask if that is supposed to work.

Reply:
Either way, we can not assume that every class that comes here will have a serializer attached to it (nor be a Kotlin class)

If there is no serializer defined, or just based on types, we can fall back on a registry that we can keep as a map of Type (Class, Class name, etc) and a Serializer. That map could be configurable so someone might add support for classes without `@Serializable` like Java Dates, _and_ we can pre-populate it with AF serializers configuration. Or keep the AF serializers in a separate config, but that's a finer detail.

[This](https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#passing-a-serializer-manually) and 
[this](https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#passing-a-serializer-manually) is what I mean, as we need to pass the serializer explicitly in the very broad generic setting anyway, we might as well keep it somewhere in a config map, similar as to how this cache is doing but is just caching _all_ serializers.